### PR TITLE
fix: make codegen more robust

### PR DESCRIPTION
**Pull request by divyenduz**
This PR does the following:-

1. Fail fast on any errors and notify the user about a potential fix

2. Refactor the logic to locate `apollo-codegen` binary to support `typegen`

3. Provide a default input value to `typegen`

4. Downgrade `npm-run` to get around the regression on windows

I am testing this one one windows after which we can merge this and hopefully make our codegen users happy. Thanks!

### CODE REVIEW

File: src/cmds/codegen/codegen.ts [hunk 1]
```diff
@@ -128,27 +127,41 @@ export class Codegen {
         }
 
         if (generator === 'typegen') {
+
+          if (!output.typings || output.typings === '') {
+            throw new Error("Please provide output.typings path to use typegen")
```
Comment (marktani) [inline src/cmds/codegen/codegen.ts]: Can we make this error message actionable? How can I usually fix this problem?

File: src/cmds/codegen/codegen.ts [hunk 2]
```diff
@@ -128,27 +127,41 @@ export class Codegen {
...
             output.typings,
             '--target',
             language,
           ]
+
           const child = crossSpawn.sync(binPath, args)
+        
+          if (child.error) {
+            if (child.error.message === `spawnSync apollo-codegen ENOENT`) {
+              throw new Error(`Generator apollo-codegen is not installed.`)
```
Comment (marktani) [inline src/cmds/codegen/codegen.ts]: Can we make this error message actionable? How can I usually fix this problem?

File: src/cmds/codegen/codegen.ts [hunk 3]
```diff
@@ -158,16 +171,31 @@ export class Codegen {
           // fs.unlinkSync(tmpSchemaPath)
         } else {
           const args = ['--input', inputSchemaPath, '--language', language]
+
+          if (!output.binding || output.binding === '' && !output.typeDefs || output.typeDefs === '') {
+            throw new Error("Please provide either output.binding or output.typeDefs to use this generator")
```
Comment (marktani) [inline src/cmds/codegen/codegen.ts]: Can we make this error message actionable? How can I usually fix this problem?

File: src/cmds/codegen/codegen.ts (hunk 2 above)
Comment (divyenduz) [reply src/cmds/codegen/codegen.ts]: Ideally, we will not hit this at all with this new code!

File: src/cmds/codegen/codegen.ts (hunk 1 above)
Comment (divyenduz) [reply src/cmds/codegen/codegen.ts]: `Please provide output.typings path in graphql config to use typegen` - does this sound better? I basically want to notify the user that they need to provide the `typings` option in the `codegen` extension.

**Comment by divyenduz** [comment]
Changes in terms of error message are done as suggested by @marktani 

Testing on windows is in progress.

**Comment by divyenduz** [comment]
Not ready for windows yet, one of our `sub-deps` is broken for windows - https://github.com/timoxley/npm-run/issues/19

Pondering a solution.

**Comment by divyenduz** [comment]
Current solution is to downgrade `npm-run` to 4.x. I can confirm that it works on windows now, I also added `npm-run` to `ignoreDeps` in `renovate.json` so that we don't accidentally update it. 

This is ready to be merged now. Please let me know if any changes are needed. Thanks!

**Comment by ntziolis** [comment]
Great job tracking this down, any chance to also downgrade npm run for prisma itself to make sure `prisma deploy` actually creates the post deployment files?

**Comment by divyenduz** [comment]
@ntziolis : I am trying to replicate the `prisma deploy` not calling `hooks` issue on windows and will fix that after successful reproduction.

**Pull request merged**